core/src/p2p.rs

```diff
@@ -0,0 +1,571 @@
+                None => {
+                    println!("P2P post failed not connected.");
+                    continue;
+                },
+            };
+            println!("POST!");
+            let encrypted = crypto.encrypt(encoded.clone()).expect("We should always be able to encrypt.");
+            let write1 = stream.write_all(&(encrypted.len() as u32).to_le_bytes());
+            let write2 = stream.write_all(&encrypted);
+
+            if write1.is_err() || write2.is_err() {
+                to_disconnect_keys.push(public_key);
+            }
+        }
+
+        for key in to_disconnect_keys {
+            println!("Disconnected during post from {}.", key);
+            connected_to_peers.remove(&key).expect("Peer to disconnect must have been in the hashmap.");
+        }
+    }
+
+    pub fn get_connected_to_peer_keys(&self) -> Vec<PublicKey> {
+        let mut keys = Vec::new();
+
+        for (public_key, (stream, crypto)) in self.connected_to_peers.lock().unwrap().iter_mut() {
+            if stream.write_all(&0_u32.to_le_bytes()).is_err() { // This has to be done in order to detect broken pipes.
+                continue;
+            }
+            keys.push(public_key.clone());
+        }
+        keys
+    }
+
+    pub fn update_peer_target(&self, new_target: &[PeerId]) {
+        let mut target = self.connect_peer_target.lock().unwrap();
+        target.clear();
+        target.extend_from_slice(new_target);
+        if let Some(index) = target.iter().position(|peer_id| peer_id.public_key == self.public_key) {
+            target.remove(index);
+        }
+    }
+}
+
+pub fn start_read_loop(p2p: Arc<P2PSystem>, sumeragi: Arc<Sumeragi>, block_sync: Arc<BlockSynchronizer>) -> ThreadHandler {
+    // Oneshot channel to allow forcefully stopping the thread.
+    let (shutdown_sender, shutdown_receiver) = tokio::sync::oneshot::channel();
+
+    let thread_handle = std::thread::spawn(move || {
+        p2p_read_loop(&p2p, shutdown_receiver, &sumeragi, &block_sync);
+    });
+
+    let shutdown = move || {
+        let _result = shutdown_sender.send(());
+    };
+
+    ThreadHandler::new(Box::new(shutdown), thread_handle)
+}
+
+fn p2p_read_loop(
+    p2p: &P2PSystem,
+    mut shutdown_receiver: tokio::sync::oneshot::Receiver<()>,
+    sumeragi: &Sumeragi,
+    block_sync: &BlockSynchronizer,
+) {
+    loop {
+        // We have no obligations to network delivery so we simply exit on shutdown signal.
+        if shutdown_receiver.try_recv().is_ok() {
+            info!("P2P read thread is being shut down");
+            return;
+        }
+        std::thread::sleep(Duration::from_micros(2500));
+
+        // For protocol violators.
+        let mut to_disconnect_keys = Vec::<PublicKey>::new();
+
+        let target = p2p.connect_peer_target.lock().unwrap().clone();
+        let mut connected_to_peers = p2p.connected_to_peers.lock().unwrap();
+        for (public_key, (stream, crypto)) in connected_to_peers.iter_mut() {
```

Comment:
Notice that in this *function*, the mantra: 
```rust 
if something.is_err() {
	to_disconnect_keys.push(public_key.clone();
	continue;
}
``` 
is repeated. The overwhelming redundancy of this statement should signal to you that in fact a custom error type would have been helpful. 

A much better solution would have been to put the entire body of the `for` loop inside a function. Forward the errors up that function with the `?` operator, and then do a single `if let Err(_)` at the top level. In the only two cases where you short-circuit without adding the key to the disconnected keys you should return `Ok`, and wrap the `else` block of the `if let` wrap the remainder of the function.